ipa-core/src/cli/crypto.rs

```diff
@@ -0,0 +1,359 @@
+        }
+    }
+
+    Ok(())
+}
+
+async fn build_hpke_registry(
+    private_key_file: PathBuf,
+) -> Result<KeyRegistry<PrivateKeyOnly>, BoxError> {
+    let mk_encryption = Some(HpkeServerConfig::File { private_key_file });
+    let key_registry = hpke_registry(mk_encryption.as_ref()).await?;
+    Ok(key_registry)
+}
+
+struct DecryptedReports {
+    reader: BufReader<File>,
+    key_registry: KeyRegistry<PrivateKeyOnly>,
+}
+
+impl DecryptedReports {
+    fn new(filename: &PathBuf, key_registry: KeyRegistry<PrivateKeyOnly>) -> Self {
+        let file = File::open(filename)
+            .unwrap_or_else(|e| panic!("unable to open file {filename:?}. {e}"));
+        let reader = BufReader::new(file);
+        Self {
+            reader,
+            key_registry,
+        }
+    }
+}
+
+impl Iterator for DecryptedReports {
+    type Item = OprfReport<BA8, BA3, BA20>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut line = String::new();
+        if self.reader.read_line(&mut line).unwrap() > 0 {
+            let encrypted_report_bytes = hex::decode(line.trim()).unwrap();
+            let enc_report =
+                EncryptedOprfReport::from_bytes(encrypted_report_bytes.as_slice()).unwrap();
+            let dec_report: OprfReport<BA8, BA3, BA20> =
+                enc_report.decrypt(&self.key_registry).unwrap();
+            Some(dec_report)
+        } else {
+            None
+        }
+    }
+}
+
+/// # Panics
+// if input files or private_keys are not correctly formatted
+/// # Errors
+/// if it cannot open the files
+pub async fn decrypt_and_reconstruct(args: DecryptArgs) -> Result<(), BoxError> {
+    let key_registry1 = build_hpke_registry(args.mk_private_key1).await?;
+    let key_registry2 = build_hpke_registry(args.mk_private_key2).await?;
+    let key_registry3 = build_hpke_registry(args.mk_private_key3).await?;
+    let decrypted_reports1 = DecryptedReports::new(&args.input_file1, key_registry1);
+    let decrypted_reports2 = DecryptedReports::new(&args.input_file2, key_registry2);
+    let decrypted_reports3 = DecryptedReports::new(&args.input_file3, key_registry3);
+
+    let mut writer: Box<dyn Write> = Box::new(
+        OpenOptions::new()
+            .write(true)
+            .create_new(true)
+            .open(args.output_file)?,
+    );
+
+    for (dec_report1, (dec_report2, dec_report3)) in
+        decrypted_reports1.zip(decrypted_reports2.zip(decrypted_reports3))
+    {
+        let timestamp = [
+            dec_report1.timestamp,
+            dec_report2.timestamp,
+            dec_report3.timestamp,
+        ]
+        .reconstruct()
+        .as_u128();
+
+        let match_key = [
+            dec_report1.match_key,
+            dec_report2.match_key,
+            dec_report3.match_key,
+        ]
+        .reconstruct()
+        .as_u128();
+
+        assert_eq!(dec_report1.event_type, dec_report2.event_type);
+        assert_eq!(dec_report2.event_type, dec_report3.event_type);
```

Comment:
```suggestion
```

you don't need to validate, `reconstruct` will [panic](https://github.com/private-attribution/ipa/blob/7e894c36a453d7918829c28da2ccddfab6553f19/ipa-core/src/test_fixture/sharing.rs#L55) if shares don't match.

Reply:
these don't come from `reconstruct` though

Reply:
adding a comment in the code to make it more clear why this is here.